### DISCUSSION TITLE
blog_posts/minimal_agx_implementation

### DISCUSSION THREAD

**Discussion by giscus** [MissingDeadlines/iolite#12] [Blog Posts]
# blog_posts/minimal_agx_implementation

Contains a minimal implementation of Troy Sobotka's AgX display transform, which is easy to drop into any existing codebase - especially for testing purposes.

https://iolite-engine.com/blog_posts/minimal_agx_implementation

**Comment by Opioid** [comment]
Thanks for this code! How did you compute the matrices? I guess linear sRGB space is assumed? I am interested in re-deriving the inset matrix so that it works for AcesCG.

**Reply by Opioid** [reply]
Oh. The matrix is part of the original repo. Am I right, that the input should be in sRGB space already?

**Reply by Opioid** [reply]
Also, what is the purpose of the offset value in agxLook()?

**Reply by begla** [reply]
The "offset" is part of the ASC-CDL transform and is left there for completeness. The default AgX look transforms don’t use it.

https://en.m.wikipedia.org/wiki/ASC_CDL

Precisely, the input is expected to be linear tristimulus with Rec. 709 primary chromaticities, often referred to as "linear sRGB". Same for the output (after `agxEotf` in the code), and you are required to apply a fitting EOTF before outputting the result.

**Reply by Opioid** [reply]
Thanks for the link!

I'm a bit confused about the EOTF part. If I understand you correctly the output is in "linear sRGB". But then we proceed to do "pow(val, vec3(2.2))". Or we could also use the piecewise "toLinear()" function present in the shadertoy, correct? But isn't that converting linear to linear?

In the shadertoy the pow(x, 2.2) is commented out. Is that because the conversion is handled by the hw-rendertarget?

My current pipeline looks roughly like this: writePngToDisk(linearToGamma(tonemap(x))). So if tonemap is main() from here, the pow(x, 2.2) part is basically cancelled out. But the results look reasonable. Leaving pow(x, 2.2) out, like in the shadertoy, leads to a pretty washed out image.

**Reply by begla** [reply]
You're right that it's a little confusing.

I quickly rechecked the official OCIO config, and your interpretation and usage are correct. The immediate output ("AgX Base" in the config file) is directly interpreted as non-linear Rec. 709 with a 2.2 gamma curve applied to it. I was referring to the output after the `agxEotf` as being linear since I did not "bake" this part.

For me, it's natural to stay linear in the shader and to use sRGB RTs; hence I added the linearization in the `agxEotf` function. So, if you're not using a sRGB render target, like it's the case for the shadertoy, the shortcut is, as you pointed out, to comment the line.

**Reply by Opioid** [reply]
Thanks a lot for clarifying!

**Reply by Luke25361** [reply]
This is fantastic, thanks so much for sharing 😄

I also have a question about the EOTF part. Is there a reason the linearization happens _after_ `agx_mat_inv` has been applied? I ask only because I noticed it happening _before_ this step instead in [MrLixm's implementation](https://github.com/MrLixm/AgXc/blob/4b48cef277ddae778f1f6327eb0c31396e853191/obs/obs-script/AgX.hlsl#L176-L177), and I'm not yet familiar enough with OCIO to fully understand what's going on in the original config file.
There's definitely a (subtle-ish) difference in output when doing the linearization before vs after, though I'm struggling to describe it...it's like the dark colours get darker and/or more saturated...

**Reply by begla** [reply]
The application of the outset, as described in the post, was recommended and inserted by Troy, the original author of AgX, after reviewing my code. The last time I checked, it was not even part of the "official" config file yet.

**Reply by Luke25361** [reply]
Gotcha, am just now realising that that's what was meant by "full round trip"
Thanks again for sharing this :)

**Reply by kofa73** [reply]
Hi, I'm late to the party, and also just a beginner at image/colour processing. Thanks for providing this implementation.
Like others before me, I'm confused about the outsetting done before linearisation. I wonder if I misunderstand the OCIO config from Troy Sobotka:
https://github.com/sobotka/SB2383-Configuration/blob/main/config.ocio#L122-L136 - there, the sequence of operations is:
```
        - !<ColorSpaceTransform> {src: Linear BT.709, dst: AgX Log (SB2383)}
        - !<FileTransform> {src: AgX_Default_Contrast.spi1d}
        - !<ExponentTransform> {value: 2.2}
        - !<MatrixTransform> {matrix: [...], direction: inverse}
        - !<ExponentTransform> {value: 2.2, direction: inverse}
```
If I understand that right, the power function with exponent 2.2 is applied **before** the inverse matrix.

What did I miss?

**Reply by allenwp** [reply]
> If I understand that right, the power function with exponent 2.2 is applied **before** the inverse matrix.

I believe you're correct in how you have read the ocio file for the SB2383 configuration of AgX; this is supported by the ordering shown in the Resolve DCTL files for [Camera-AgX](https://github.com/sobotka/AgX-Resolve/blob/40dce3acfaf8ae0bdf9a540ea7433a43c4a75153/Camera-AgX.dctl#L288-L290) and [Kraken-AgX](https://github.com/sobotka/AgX-Resolve/blob/40dce3acfaf8ae0bdf9a540ea7433a43c4a75153/Kraken-AgX.dctl#L291-L293).

I talked briefly with Troy, the original AgX author, about the difference in performing the power function linearization step before or after the outset matrix. Simply put, both approaches are equally correct in terms of [goals that the original AgX set out to achieve](https://github.com/sobotka/AgX-S2O3?tab=readme-ov-file).

You'll notice that [other configurations of AgX](https://github.com/JuanPabloZambrano/DCTL/blob/954de6c3be2a588a00b0ddf24ba1f424bbd7b7ec/AgxDRT/Jp-AgxDRT.dctl#L143-L149) may perform linearization after the outset matrix.

All this to say, you should try linearization before and after the outset matrix and pick whichever one you prefer the look of. Or create your own AgX configuration based on the [SB2383 generator code](https://github.com/sobotka/SB2383-Configuration-Generation) or others.

**Reply by sobotka** [reply]
When dealing with the basis vectors of the colourimetric stimuli model, the goal is to apply the matrix transforms on discrete wattage samples.

In the outline @kofa73 pasted above, we can see the transform to a normalized log, and then the “curve” is applied. That curve includes the code value encoding in it; we could display the state if we wanted to, simply by shipping it off to the display, immediately after that file transform 1D LUT of the baked formula, and it would “look close-ish” because the display in turn would roll the code values passed to it out through the 2.2 power function EOTF.

So the `ExponentTransform` applies the EOTF to recover the discrete scalar wattages, prior to rolling through the 3x3 colourimetric stimuli matrix. The “by the book” approach for transforming colourimetry.

The final `ExponentTransform` then re-encodes the discrete scalar wattages back to the display encoding.

Hope this help. Let me know if I have made matters worse. 🤣

**Reply by kofa73** [reply]
Thanks to both of you.
One more thing:
```
vec3 agxLook(vec3 val) {
  const vec3 lw = vec3(0.2126, 0.7152, 0.0722);
  float luma = dot(val, lw);
  [...]
  
  // ASC CDL
  val = pow(val * slope + offset, power);
  return luma + sat * (val - luma);
}
```
This means that if saturation is 0, all the other controls (slope, offset, power) are ignored, and we fall back to merely returning the luma of the **input**. https://docs.acescentral.com/specifications/acescct/#appendix-a-application-of-asc-cdl-parameters-to-acescct-image-data does not specify whether the control is to be applied to `ACES_cct_in` or `ACES_cct_out`, but it seems logical it's the output. I'm not familiar with the details, but I think what is implied here by the pipeline-like arrangement:

https://github.com/AcademySoftwareFoundation/OpenColorIO/blob/e6622965f3a71e3fa6ff39e888eca0b3b4f283f2/src/OpenColorIO/ops/cdl/CDLOp.cpp#L227-L237

**Reply by sobotka** [reply]
I’d exert extreme caution on these ideas behind “saturation”. For starters those weights are BT.1886 / BT.709, and applying them against other RGB can and will cause cognitive weirdness.

The canonical formula is actually just a simple linear interpolation between the RGB wattage and the `luminance(RGB)`. It’s odd that the formula listed doesn’t seem to do that? Normally it is broadly `(RGB * amount) + ((100% - amount) * luminance(RGB))`. Not sure why the formula pasted isn’t doing the more traditional “saturation”, and I suspect that’s a bug.

Regardless, this formula will potentially thrash the pictorial surface. While no one has “solved” this problem, we can make a few conjectures:
1. “saturation” is cognitively “assigned” to forms, based on segmentation-decomposition, within global to local frameworks.
2. Within a *local* framework, a form will often be “enmeshed” within sub-forms. For example, if we imagine a sphere coated with a semi-matte clear coat, the clear coat will additively reflect other energy, and the excitation purity of the base coat will be reduced. Our cognitive machinery appears to segment and decompose this “layer” from the “base”.
3. The formula above has no care about local frameworks. As such, it will likely shred most pictures by shifting the gradients such that segmentation-decomposition cognitive mechanisms will struggle. Following our example above, we can imagine the semi-matte layer being “eroded” in a peculiar manner.
4. I am skeptical that a linear interpolation from `chrominance + luminance -> luminance` is prudent for the above reasons. If our segmentation-decomposition engine exists and leans on dominant energy, it follows that as we reduce chrominance, the luminance signal should be gained toward the dominant energy.

A simple example follows. Notice how the cognized position of the “sphere-like” form leads to a cognition of the form “drawing nearer” and “getting squished against a near plane” as the saturation control increases.

https://github.com/user-attachments/assets/18634257-6090-42fa-9c1c-9509fb219f2a

**Reply by begla** [reply]
To clear this up, both expressions are equivalent. The one I used saves one multiplication by rearranging the more commonly known equation (linear interpolation):

```math
\begin{align}
l &= \text{luminance(rgb)}\\
s &= \text{user-defined saturation factor in} \left[0,1\right]\\
v &= \text{r, g, or b of rgb}\\\\
l + s (v - l) &\equiv s v + (1 - s) l\\\\
&\equiv l + s (v - l) && \text{distribute} \\
&\equiv l + sv - sl && \text{rearrange} \\
&\equiv sv + l - sl \equiv sv + (l - sl) && \text{factor out l}\\
&\equiv sv + (1 - s) l
\end{align}
```

Sorry for causing confusion here! Thanks, @sobotka, for jumping in and providing detailed explanations. 🙂

**Reply by kofa73** [reply]
I just meant that 

` float luma = dot(val, lw);`

should probably come after

`val = pow(val * slope + offset, power);`

```
const vec3 lw = ...
  
  [...]
  
// ASC CDL
val = pow(val * slope + offset, pow
float luma = dot(val, lw);
return luma + sat * (val - luma);
```

**Reply by sobotka** [reply]
Agree with @kofa73, which was my point too; the creative adjustment would be dependent on the linear interpolation, which seems off. It would seem to be linearly interpolating between the creative adjustment and the luminance?

**Reply by begla** [reply]
I agree!

I've adjusted the blog post and shadertoy to compute the luminance value for the saturation adjustment based on the results of the ASC CDL transform.

Thanks, guys, for bringing this to my attention. 👍

**Comment by hybridherbst** [comment]
Thanks for providing an easy-to-follow implementation!
I'd have two questions:
- Is there a reference viewer somewhere to compare results to?
- When using a toneMappingExposure multiplier, would that be just a multiplier before applying agx or would it be elsewhere?
  ```
        color *= toneMappingExposure; // apply an exposure multiplier
        color = agx(color);
  ```

**Reply by begla** [reply]
That's a perfectly valid way to apply exposure!

Regarding a reference render, I used the official OCIO config and Blender to validate the results, which is also relatively easy to set up.

Troy's testing imagery collection provides some useful sample images for doing any form of comparison/validation:

https://github.com/sobotka/Testing_Imagery

**Reply by hybridherbst** [reply]
Thank you! Do you know if the Blender 4.x implementation of AgX (which seems to be the default tonemapping by now) is different to your approach?

**Reply by begla** [reply]
I'm not aware of any changes to the reference implementation, but I also have not checked the actual 4.0 implementation in detail yet.

**Reply by donmccurdy** [reply]
From discussion during development for three.js, one difference is the "guard rail", which has some effects you can see here:

https://github.com/mrdoob/three.js/pull/27366#issuecomment-1862040027

In our case, it didn't seem feasible to include that in a realtime implementation, and the differences are acceptable. Blender [screenshots were generated with a custom bash script and OCIO](https://gist.github.com/donmccurdy/ca1d7cbb107435b40fa510f783f91ee8).

**Comment by Calinou** [comment]
Hi, thanks for this code! It seems to work quite well in my [Godot prototype](https://github.com/godotengine/godot-proposals/discussions/7545#discussioncomment-7988168) :slightly_smiling_face:

I have two questions:

- How is the whitepoint defined? Other tonemapping methods like Filmic and ACES have an adjustable whitepoint, so you can choose beyond which point highlights will be blown out (increasing whitepoint will make the scene a bit darker as a result). I haven't found an obvious way to do this with the AgX implementation provided here.
- Could you [add a license](https://choosealicense.com/) to the code, please? This ensures we can use it in projects legally. For context, I'm interested in [integrating this in Godot](https://github.com/godotengine/godot-proposals/discussions/7545) to provide AgX tonemapping support. Godot is MIT-licensed, so the code would need to be under MIT or a compatible license.

**Reply by begla** [reply]
That’s great! 😊

I’ve added an MIT license header to my implementation on the website and a note that all values used for deriving the implementation are sourced from Troy’s OCIO config file.

Regarding the white point:

It should be possible to implement the standard parameters by adjusting the min/max EV-based normalization (black and white points) after moving to log2 space and by scaling the open domain values right beforehand (exposure). The values I used are based on the values in the mentioned config file.

You can find inspiration for tweaking AgX by looking at Troy’s config generator. The functions handling the open domain to log2 space conversion should be the right place to start looking.

https://github.com/sobotka/AgX-S2O3/blob/main/AgX.py

**Reply by begla** [reply]
As a reference, here's an option to provide a tweakable black/white point in the open domain for the given implementation (values chosen to match the inverse of the default values):

```glsl
const float black_point = 0.00017578;
const float white_point = 16.2917;
const float min_ev = log2(black_point);
const float max_ev = log2(white_point);
```

**Reply by Calinou** [reply]
Thanks a ton! I've opened a pull request for Godot: https://github.com/godotengine/godot/pull/87260

**Reply by begla** [reply]
@Calinou Very cool, glad it worked out! 🙌

**Comment by donmccurdy** [comment]
Thank you for sharing this work! 

During review of an implementation in the three.js open source library, a question was raised about the luminance coefficients: 0.2126, 0.7152, 0.0722. These appear to be the standard coefficients for BT.709 primaries, but (to my understanding) we're working in a log space derived from Rec.2020 when applying the AgX looks. Do you know whether that's correct, and (if so) why that would be?

**Reply by begla** [reply]
At the time, I tried to match the output of the OCIO configuration file Troy initially provided.

https://github.com/sobotka/AgX/blob/main/config.ocio

I remember doing some backtracking regarding the ASC-CDL transform ("CDLTransform" in the config file) used for implementing the looks. The ASC-CDL does not incorporate a saturation parameter in most openly available documentation. I also suspected some context-sensitive behavior according to the involved primaries and color spaces.

But, as far as I can tell, the luma coefficients in the "CDLTransform" of OCIO are fixed and match the ones I ended up using:

https://github.com/AcademySoftwareFoundation/OpenColorIO/blob/aadf595946ed8a27d69ad0e338d0a8b779465707/src/OpenColorIO/transforms/CDLTransform.cpp#L302

**Reply by donmccurdy** [reply]
Thank you @begla! I haven't been able to find documentation for ASC-CDL v1.2 about this either, but that sounds like a good explanation of the choice to me. :)

**Comment by nxrighthere** [comment]
Thank you for the code! Here's the implementation for UE 5.3 https://gist.github.com/nxrighthere/eb208dae8b66dbe452af223f276e46cc

**Comment by bWFuanVzYWth** [comment]
```
vec3 saturate(vec3 v) {
    return clamp(v, 0.0, 1.0);
}

vec3 agx_curve3(vec3 v) {
    const float threshold = 0.6060606060606061;
    const float a_up = 69.86278913545539;
    const float a_down = 59.507875;
    const float b_up = 13.0 / 4.0;
    const float b_down = 3.0 / 1.0;
    const float c_up = -4.0 / 13.0;
    const float c_down = -1.0 / 3.0;

    vec3 mask = step(v, vec3(threshold));
    vec3 a = a_up + (a_down - a_up) * mask;
    vec3 b = b_up + (b_down - b_up) * mask;
    vec3 c = c_up + (c_down - c_up) * mask;
    return 0.5 + (((-2.0 * threshold)) + 2.0 * v) * pow(1.0 + a * pow(abs(v - threshold), b), c);
}

vec3 agx_tonemapping(vec3 /*Linear BT.709*/ci) {
    const float min_ev = -12.473931188332413;
    const float max_ev = 4.026068811667588;
    const float dynamic_range = max_ev - min_ev;

    const mat3 agx_mat = mat3(0.8424010709504686, 0.04240107095046854, 0.04240107095046854, 0.07843650156180276, 0.8784365015618028, 0.07843650156180276, 0.0791624274877287, 0.0791624274877287, 0.8791624274877287);
    const mat3 agx_mat_inv = mat3(1.1969986613119143, -0.053001338688085674, -0.053001338688085674, -0.09804562695225345, 1.1519543730477466, -0.09804562695225345, -0.09895303435966087, -0.09895303435966087, 1.151046965640339);

    // Input transform (inset)
    ci = agx_mat * ci;

    // Apply sigmoid function
    vec3 ct = saturate(log2(ci) * (1.0 / dynamic_range) - (min_ev / dynamic_range));
    vec3 co = agx_curve3(ct);

    // Inverse input transform (outset)
    co = agx_mat_inv * co;

    return /*BT.709 (NOT linear)*/co;
}
```
This is not a fitting, but the exact solution I obtained through algebraic simplification.  
https://github.com/bWFuanVzYWth/AgX

**Comment by Ivan-Veselov** [comment]
But what is the point of `pow(val, vec3(2.2));`? The comment says that its for linearization, but color was never gamma-encoded in the first place. Plus EOTF is usually a part of hardware, and all rendering APIs and engines expect encoded color values as outputs not decoded display linear values. I can only assume that AgX machinery somewhere inside encodes color with an equivalent of `pow(val, vec3(1.0 / 2.2));`. But that looks strange, cause why whould one make it a part of a tonemapping operator when there are other color spaces besides sRGB, like Rec709 which has the same primaries and white point but different trasfer function (approximately 2.4 power)? Could someone help me out understanding this?

**Reply by Ivan-Veselov** [reply]
Saw the discussion about this in the replies to the first comment after posting my. So, looks like indeed AgX somewhere inbetween kind of applies power function. Still weird though.

**Comment by allenwp** [comment]
Your polynomial is a pretty great equivalent to the original curve! I'm looking to make a polynomial approximation of the Blender AgX sigmoid curve, which is similar to this one. I've been trying with different sample sizes as input to a calculator, with or without some extra padding... and a lot of trial and error! What tools/calculators/process did you use to get this formula?

**Reply by allenwp** [reply]
I managed to get a curve that was just like I wanted by using Excel trendline and adding a couple of data points before and after the range that I cared most about to finesse the curve to have good intersections where I cared most. Excel only seems to generate up to a 6th order polynomial regression, but that was exactly what I needed, similar to this post. After getting the polynomial to behave the way I wanted, I was able to set the "intercept" to 0.0, where I wanted it, which removed the addition of a negligible first coefficient. Setting the intercept to 0.0 before tweaking the curve with those extra data points caused issues, so it was important to tweak the curve first so that intersection was already happening very very close to 0.0.
